bolt12/offer.go

```diff
@@ -0,0 +1,273 @@
+package bolt12
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"net/http"
+	"net/url"
+	"regexp"
+	"time"
+
+	"log"
+
+	"github.com/breez/breez-lnurl/dns"
+	"github.com/breez/breez-lnurl/persist"
+	"github.com/breez/lspd/lightning"
+	"github.com/gorilla/mux"
+)
+
+const (
+	// https://datatracker.ietf.org/doc/html/rfc5322#section-3.4.1
+	// https://stackoverflow.com/a/201378
+	USERNAME_VALIDATION_REGEX = "^(?:[a-zA-Z0-9!#$%&'*+\\/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+\\/=?^_`{|}~-]+)*|\"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*\")$"
+	// https://www.rfc-editor.org/errata/eid1690
+	MAX_USERNAME_LENGTH = 64
+)
+
+type RegisterBolt12OfferRequest struct {
+	Time      int64  `json:"time"`
+	Username  string `json:"username"`
+	Offer     string `json:"offer"`
+	Signature string `json:"signature"`
+}
+
+type RegisterRecoverBolt12OfferResponse struct {
+	BIP353Address string `json:"bip353_address"`
+}
+
+func (w *RegisterBolt12OfferRequest) Verify(pubkey string) error {
+	if math.Abs(float64(time.Now().Unix()-w.Time)) > 30 {
```

Comment:
Could we have a constant for the 30 seconds? Also, I'm wondering if 30 seconds may be too strict. Not sure how likely are clock differences above 30 seconds so this may be fine.